lib/IRGen/LocalTypeData.cpp

```diff
@@ -408,26 +408,59 @@ static void maybeEmitDebugInfoForLocalTypeData(IRGenFunction &IGF,
   // functions that were inlined into transparent functions. Correct would be to
   // check which instruction requests the type metadata and see whether its
   // inlined function is transparent.
-  auto * DS = IGF.getDebugScope();
+  auto *DS = IGF.getDebugScope();
   if (DS && DS->getInlinedFunction() &&
       DS->getInlinedFunction()->isTransparent())
     return;
-  
-  // Only for formal type metadata.
-  if (key.Kind != LocalTypeDataKind::forFormalTypeMetadata())
+  // For formal type metadata and witness tables.
+  ProtocolDecl *proto = nullptr;
+
+  if (key.Kind.isAbstractProtocolConformance())
+    proto = key.Kind.getAbstractProtocolConformance();
+  else if (key.Kind.isConcreteProtocolConformance())
+    proto = key.Kind.getConcreteProtocolConformance()->getProtocol();
+  else if (key.Kind != LocalTypeDataKind::forFormalTypeMetadata())
     return;
 
   // Only for archetypes, and not for opened/opaque archetypes.
   auto type = dyn_cast<ArchetypeType>(key.Type);
   if (!type)
     return;
-  if (!type->isRoot())
+  if (!type->isRoot() && !proto)
     return;
   if (!isa<PrimaryArchetypeType>(type) && !isa<PackArchetypeType>(type))
     return;
 
-  auto *typeParam = type->getInterfaceType()->castTo<GenericTypeParamType>();
-  auto name = typeParam->getName().str();
+  auto interfaceType = type->getInterfaceType();
+  llvm::SmallString<16> name;
+  llvm::SmallString<16> displayName;
+  if (auto DMT =
+          llvm::dyn_cast<DependentMemberType>(interfaceType.getPointer())) {
+    std::function<void(DependentMemberType *)> visitDependentMembers =
+        [&](DependentMemberType *member) {
+          if (member == nullptr)
+            return;
+          if (auto *parent =
+                  llvm::dyn_cast<DependentMemberType>(member->getBase())) {
+            visitDependentMembers(parent);
+            name.append("$");
+            displayName.append(".");
+          }
+          name.append(member->getName().str());
+          displayName.append(member->getName().str());
+        };
+    name.append(DMT->getRootGenericParam()->getCanonicalName().str());
```

Comment:
Why not just print the generic param as the base case of the recursion? (where member->getBase() is not another DMT)